### DISCUSSION TITLE
Using @astrojs/node middleware in the development environment

### DISCUSSION THREAD

**Discussion by jcalfee** [withastro/roadmap#507] [Proposal]
# Summary

We are going to need a way to use [@astrojs/node](https://docs.astro.build/en/guides/integrations-guide/node) middleware in the development environment (`astro dev`).

# Background & Motivation

The `@astrojs/node` package allows for the integration of middleware such as expressjs in production.  The [middleware instructions](https://docs.astro.build/en/guides/integrations-guide/node/#middleware) addresses production by (by running an entry-point in `./dist`) but does not address hot updates during development.

What might be the most efficient salable long term solution for this?  Hot deploy and HMR are pretty common these days, it greatly improves development times and productivity.  It would be nice to have this available in Astro given the variety of components that can be deployed.

# Goals

- Efficient hot-deploy (restart if necessary) any changes to middleware source code files.
- Allow for both Astro and the middleware to run in the same NodeJS process.
- Simultaneously use the standard hot-deploy that `astro dev` implements now.
- Should work with client-side component hydration.

**Comment by jcalfee** [comment]
What about just simply exporting the request response handler?  Middleware is usually very simple, a `nodemon` can be enough to implement a hotreload like setup.  The astro build exports a handler that can be used to hook into the request response chain.  Why can't the astro dev export this handler too?  Astro dev starts pretty fast, I guess it could just restart the dev server.

**Reply by jcalfee** [reply]
This is blocking me from starting with astro.  I can't use astro because have no way to get a consistent environment between dev and production.

**Reply by matthewp** [reply]
You might look at https://github.com/matthewp/astro-fastify for inspiration, as that adapter does run Fastify plugins during development mode. I think a similar approach could be taken for the Node middleware mode adapter.

**Reply by jcalfee** [reply]
Wow .. that  is pretty intense.  I think I'll be using just what you have and slowly adapting it as I get my head around it.  Thank you!  This is a big head start.

**Reply by elee1766** [reply]
ok i tried my hand at this because i was playing around with this idea

https://github.com/gfx-labs/astro-express
